### PR TITLE
Add warning for including empty segments or files in optional mode

### LINKED ISSUE #2181: Log warning if the include fragment is empty
> Let's put up an issue for discussion on the suggestion that if the fragment is empty, perhaps we can consider outputting a warning since that could be a case of incorrect usage.

_Originally posted by @tlylt in https://github.com/MarkBind/markbind/issues/2180#issuecomment-1436518027_

### PR DESCRIPTION
**Pull request by Tim-Siu**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

**Overview of changes:**
Related to #2180
Resolve #2181 
Add warning message in the case that empty files or segments are included in the `<include .... optional/>` case.

**Anything you'd like to highlight/discuss:**

**Testing instructions:**

Minimal reproduction:
```
<div id="empty-segment">
<include src="some_path#empty" optional/>
</div>

<div id="empty">
<include src="some_path" optional/>
</div>
```
The log warning expected:
```
warn: Optional segment '#empty' not found in file: ...
Empty reference in ...
warn: Optional file not found: ...
Empty reference in ...
```

**Proposed commit message: (wrap lines at 72 characters)**

Add warning for including empty segments or files in optional mode
---

**Checklist:** :ballot_box_with_check:

- [] Updated the documentation for feature additions and enhancements
- [] Added tests for bug fixes or features
- [x] Linked all related issues
- [] No unrelated changes 

---

**Reviewer checklist:**

Indicate the [SEMVER](https://semver.org/) impact of the PR:
- [ ] Major (when you make incompatible API changes)
- [ ] Minor (when you add functionality in a backward compatible manner)
- [ ] Patch (when you make backward compatible bug fixes)

At the end of the review, please label the PR with the appropriate label: `r.Major`, `r.Minor`, `r.Patch`.

Breaking change release note preparation (if applicable):
- To be included in the release note for any feature that is made obsolete/breaking

> Give a brief explanation note about:
>   - what was the old feature that was made obsolete
>   - any replacement feature (if any), and
>   - how the author should modify his website to migrate from the old feature to the replacement feature (if possible).

### CODE REVIEW

**Comment by yucheng11122017** [comment]
Hi @Tim-Siu, is this the desired behaviour? 
My impression from reading the issue and the comments is that there is a need for a warning message if the include content is empty, regardless of whether the fragment is optional or not. 

In other words, 
```
<includes src="path#empty" />
<div id="empty" />
```
will log a warning regardless of whether the include is optional or not.

**Comment by yucheng11122017** [comment]
Also once again, please look at the checklist to see if you missed out on anything and check it off if you have done it.

> **Checklist:** :ballot_box_with_check:
> 
> 
> 
> - [ ] Updated the documentation for feature additions and enhancements
> - [ ] Added tests for bug fixes or features
> - [ ] Linked all related issues
> - [ ] No unrelated changes

**Comment by Tim-Siu** [comment]
> Hi @Tim-Siu, is this the desired behaviour? My impression from reading the issue and the comments is that there is a need for a warning message if the include content is empty, regardless of whether the fragment is optional or not.
> 
> In other words,
> 
> ```
> <includes src="path#empty" />
> <div id="empty" />
> ```
> 
> will log a warning regardless of whether the include is optional or not.

Hi @yucheng11122017 ,

Currently empty includes without `optional` attributes will lead to error (and will be printed to both console and rendered web pages).

**Comment by Tim-Siu** [comment]
> @Tim-Siu Also once again, please look at the checklist to see if you missed out on anything and check it off if you have done it.
> 
> > **Checklist:** ☑️
> > 
> > 
> > * [ ]  Updated the documentation for feature additions and enhancements
> > * [ ]  Added tests for bug fixes or features
> > * [ ]  Linked all related issues
> > * [ ]  No unrelated changes

Hi @yucheng, 

Regarding the checklist:
1. I will add related documentation when the desired behaviors are agreed upon.
2. To my best knowledge, testing for logging is currently not available?
3. I failed to mention a previous related PR. Thank you for the reminding.
4. I Fixed a small typo in the same code file. I will remove the fix if it is against conventional practices.

Thank you once again for your comments.

**Comment by yucheng11122017** [comment]
> > Hi @Tim-Siu, is this the desired behaviour? My impression from reading the issue and the comments is that there is a need for a warning message if the include content is empty, regardless of whether the fragment is optional or not.
> > In other words,
> > ```
> > <includes src="path#empty" />
> > <div id="empty" />
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > will log a warning regardless of whether the include is optional or not.
> 
> Hi @yucheng11122017 ,
> 
> Currently empty includes without `optional` attributes will lead to error (and will be printed to both console and rendered web pages).

Hi @Tim-Siu, I believe that the purpose of #2181 is to log when a user includes on a fragment that is empty. 

You can take a look at the [test case](https://github.com/MarkBind/markbind/pull/2180/files#diff-ede52d35aee85f528271c5f36627945c37608764f10ac0dc45aa9edad2a9b04b) added by the original PR which lead to this issue. The idea is that if there a empty fragment like <div id="empty />, there should be a log that tells the user they are including an empty fragment. 

Eg. 
In trial.md
```
<includes src="trialInclude.md#empty" />
```

In trialInclude.md
```
<div id="empty" />
```
Because the div is empty, we add a log statement for the user. 

Right now, you are logging only if the fragment does not exist and optional is enabled. Eg. 

In trial.md
```
<include src="trialInclude.md#missing" optional />
```
In trialInclude.md there is no tag with missing id. 

We don't need this behaviour when optional is enabled because the whole idea is that if optional is enabled, the fragment can be skipped. Otherwise, the user shouldn't have enabled optional in the first place.

File: packages/core/src/html/includePanelProcessor.ts [hunk 1]
```diff
@@ -269,6 +269,10 @@ export function processInclude(node: MbNode, context: Context, pageSources: Page
       logger.error(error);
 
       actualContent = cheerio.html(createErrorNode(node, error));
+    } else if (actualContentOrNull === '') {
+      const warning = `Segment '${hash}' in file: ${actualFilePath} is empty\n`
+          + `Empty reference in ${context.cwf}`;
```
Comment (yucheng11122017) [inline packages/core/src/html/includePanelProcessor.ts:274]:
```suggestion
      const warning = `Empty reference in ${context.cwf}:\n`
          + `Segment '${hash}' in file: ${actualFilePath} is empty`;
```

**Review by yucheng11122017 — COMMENTED** [review]
Hi @Tim-Siu thanks for the work! 

One clarification: is the intention also to add a logging statement if the included file is empty? 

Eg. 
In trial.md
```
<includes src="trialInclude.md" />
```

while trialInclude.md is present but the file is empty. 

This behaviour doesn't trigger a warning right now.

**Comment by yucheng11122017** [comment]
Could you also fix the fact that the `#example` in site navigation menu being empty so that the build doc doesn't fail? 

It can be similiar to the one for page nav
```
<div id="examples" class="d-none">

You can see an example of a Page Navigation Bar ==on the right side== of <a target="_blank" href="{{ baseUrl }}/userGuide/formattingContents.html">this page</a>.
</div>
```

File: packages/core/src/html/includePanelProcessor.ts [hunk 2]
```diff
@@ -66,7 +66,7 @@ function _getBoilerplateFilePath(element: MbNode, filePath: string, config: Reco
 function _getSrcFlagsAndFilePaths(element: MbNode, config: Record<string, any>) {
   const isUrl = urlUtil.isUrl(element.attribs.src);
 
-  // We do this even if the src is not a url to get the hash, if any
+  // We do this even if the src is not an url to get the hash, if any
```
Comment (itsyme) [inline packages/core/src/html/includePanelProcessor.ts:69]: 'a' is actually the correct word as url is pronounced as 'you-ar-el' which leads to the first "sound" letter being 'y'. This article explains in more detail https://owl.purdue.edu/owl/general_writing/grammar/articles_a_versus_an.html

Comment (Tim-Siu) [reply packages/core/src/html/includePanelProcessor.ts:69]: Sorry for this. It was highlighted by static analysis and I thought it was a valid typo HAHA

**Comment by Tim-Siu** [comment]
HI @yucheng11122017 

Thank you for your suggestions. I have made changes accordingly.

File: docs/userGuide/syntax/siteNavigationMenus.md [hunk 3]
```diff
@@ -34,7 +34,10 @@ You can **append the `:expanded:` to a <tooltip content="a menu item with sub me
 
 </div>
 
-<div id="examples"></div>
+<div id="examples" class="d-none">
+
+You can see an example of a Site Navigation Menu ==on the top== of <a target="_blank" href="{{ baseUrl }}/userGuide/formattingContents.html">this page</a>.
```
Comment (yucheng11122017) [inline docs/userGuide/syntax/siteNavigationMenus.md:39]: Er isn't it on the left?

**Review by yucheng11122017 — COMMENTED** [review]
Small nit

**Review by yucheng11122017 — APPROVED** [review]
LGTM!

**Pull request merged**